<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Schmutzi : Maximum a posteriori estimate of contamination for ancient DNA and endogenous mitochondrial genome inference ">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>schmutzi</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/grenaud/schmutzi">View on GitHub</a>

          <!-- <h1 id="project_title">schmutzi</h1> -->
	  <center>
	  <img src="https://github.com/grenaud/schmutzi/raw/gh-pages/images/schmutziLogo.png" alt="schmutzi" height="264" width="600"> 
	  </center>
          <h2 id="project_tagline">Maximum a posteriori estimate of contamination for ancient DNA and endogenous mitochondrial genome inference </h2>

            <section id="downloads">
          <!--              <a class="zip_download_link" href="https://github.com/grenaud/schmutzi/zipball/master">Download this project as a .zip file</a> -->
              <a class="tar_download_link" href="https://bioinf.eva.mpg.de/schmutzi/schmutzi.tar.gz">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="About" class="anchor" href="#about" aria-hidden="true"><span class="octicon octicon-link"></span></a>About.</h3>

<p>Sequencing of mitochondrial DNA from ancient samples is often used as a preliminary strategy to ensure recoverable endogenous DNA, that is DNA endogenous to sample of interest. Deamination of the molecules, a common form of damage to aDNA, can interfere with alignment and consensus calling against a reference sequence. For hominin samples, contamination from modern humans involved in the excavation and experimental process can further complicate consensus calling meaning quantification of this contamination is crucial in evaluating which ancient samples have the most potential for further downstream analyses. </p> 

<p>schmutzi is a Bayesian maximum <i>a posteriori</i> estimator of contamination for ancient hominin samples which also calls the endogenous mitochondrial consensus. 

More precisely, schmutzi is a set of programs aimed at ancient DNA data that can :

 <ul>
   <li> estimate contamination based on deamination patterns alone </li>
   <li> call a mitochondrial consensus for the endogenous genome accounting for both contamination and deamination.</li>
      <li> estimate mitochondrial contamination using the endogenous calls. By scanning a database of modern human putative contamination sources it can identify the most likely contaminant among the database records. An exhaustive set of human MTs is provided with the package.  </li>
 </ul>


<h3>
<a id="downloadinstall" class="anchor" href="#downloadinstall" aria-hidden="true"><span class="octicon octicon-link"></span></a>Download and installation</h3>

To download, either download the tar.gz file from the link on top of the page or, if you have git installed, download the latest version using:

<pre>git clone  --recursive https://github.com/grenaud/schmutzi.git</pre>

Installation instructions are provided in the README.md file

<h3>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h3>

Documentation and usage is found in the README.md file


<h3>
<a id="testdata" class="anchor" href="#testdata" aria-hidden="true"><span class="octicon octicon-link"></span></a>Test Data</h3>

There is test data provided as part of the package along with command line. See <a class="schmutzi" href="https://bioinf.eva.mpg.de/schmutzi/">https://bioinf.eva.mpg.de/schmutzi/</a> for the data used in the manuscript.

<h3>
<a id="Contact" class="anchor" href="#contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Support or Contact</h3>

<p>The program was developed and is maintained by Gabriel Renaud (<a href="https://github.com/grenaud" class="user-mention">@grenaud</a>) at the Max Planck Institute for Evolutionary Anthropology.</p>


  </body>
</html>
